### PR TITLE
Dashboard update

### PR DESCRIPTION
**Pull request by voogt**

### CODE REVIEW

File: django_project/civitas/models/view/reporter_data.py [hunk 1]
```diff
@@ -1,19 +1,25 @@
...
 
-import datetime
-
 from django.contrib.gis.db import models
 from django.db.models import Sum, Case, When, F
 from django.db.models.functions import Concat
 from django.db.models.lookups import GreaterThan, LessThan, GreaterThanOrEqual, LessThanOrEqual
+from collections import defaultdict
+from itertools import groupby
+import operator
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:10]: hi @voogt 
Please reorder everything
I think i have shared how the order of imports

File: django_project/civitas/models/view/reporter_data.py [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
 
 from civitas.models.community import Community
 
+class Subscriptable:
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:14]: I don't see it is being used
Please delete it

File: django_project/civitas/models/view/reporter_data.py [hunk 3]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
             .values('risk_level',) \
             .distinct() \
             .annotate(label=Concat('risk_level', models.Value('')), values=Sum('renewal_cost'))    
+        
+        risk_level_list = [{"risk_level":"Extreme", "index": 5}, {"risk_level":"High" , "index": 4}, {"risk_level":"Medium" , "index": 3}, {"risk_level":"Low", "index": 2}, {"risk_level":"Minimal", "index": 1}, {"risk_level":"None", "index": 0}]
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:313]: It is too long
You can change it per line
```
[
    {"risk_level": "Extreme", "index": 5},
    {"risk_level": "High", "index": 4},
    {"risk_level": "Medium", "index": 3},
    {"risk_level": "Low", "index": 2},
    {"risk_level": "Minimal", "index": 1},
    {"risk_level": "None", "index": 0}
]
```

File: django_project/civitas/models/view/reporter_data.py [hunk 4]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+            """remaining_years_renewal_system
+
+            Returns:
+                formatted list to be used correctly in table and graph
+            """
+
+            # create interval of 5 years
+            remaining_years_formatted_list = []
+            for x in range(95, -6, -5):
+                remaining_years_formatted_list.append({'remaining_years': f"{x} - {x+4}", 'risk_level': [], 'total': 0})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:325]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 5]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+                        if k.risk_level is None:
+                                risk_lev = "None"
+                        else:
+                            risk_lev = k.risk_level.strip()
+
+                        risk_level_index = [i for i,_ in enumerate(risk_level_list) if _['risk_level'] == risk_lev][0]
+
+                        if not any(d['level'] == risk_lev for d in val['risk_level']):
+                            if k.renewal_cost is not None:
+                                val["risk_level"].append({'level': risk_lev, 'values': k.renewal_cost, 'index': risk_level_list[risk_level_index]['index']})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:349]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 6]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+                                val['total'] = float(val['total']) + float(k.renewal_cost)
+                            else: 
+                                val['total'] = float(0.00)
+                                val["risk_level"].append({'level': risk_lev, 'values': 0.00, 'index': 0})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:353]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 7]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+                            if k.renewal_cost is not None:
+                                val['risk_level'][index]['values'] = float(val['risk_level'][index]['values'])+ float(k.renewal_cost)
+                                val['total'] = float(val['total']) + float(k.renewal_cost)
+
+            # add system names where value is 0 for graph to work correctly
+            for k in remaining_years_formatted_list:
+                for lev in risk_level_list:
+                    if not any(d['level'] == lev["risk_level"] for d in k["risk_level"]):
+                        risk_level_index = [i for i,_ in enumerate(risk_level_list) if _['risk_level'] == lev["risk_level"]][0]
+                        k["risk_level"].append({'level': lev["risk_level"], 'values': 0.00, 'index': risk_level_list[risk_level_index]['index']})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:365]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 8]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+            """remaining_years_renewal_system
+
+            Returns:
+                formatted list to be used correctly in graph
+            """
+            formatted_list = []   
+
+            # start with basic list
+            for risk in risk_level_list:
+                formatted_list.append({'name': risk["risk_level"], 'type': 'bar', 'x': [], 'y': []})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:397]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 9]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+                formatted list to be used correctly in table and graph
+            """
+
+            # find lowest value
+            # lowest = (min(data, key=lambda x:x['remaining_years']))
+
+            # create interval of 5 years
+            remaining_years_formatted_list = []
+            for x in range(95, -6, -5):
+                remaining_years_formatted_list.append({'remaining_years': f"{x} - {x+4}", 'asset': [], 'total': 0})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:423]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 10]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+            for k in remaining_years_formatted_list:
+                new_list = sorted(k['asset'], key=operator.itemgetter('system_name'))     
+                k['asset'] = new_list
+  
+            return reversed(remaining_years_formatted_list)
+        
+        def remaining_years_renewal_total(input_list):
+            formatted_list = []  
+            for system in system_names:
+                formatted_list.append({'name': system["system_name"], 'total': 0})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:468]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 11]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+            """remaining_years_renewal_system
+
+            Returns:
+                formatted list to be used correctly in graph
+            """
+            formatted_list = []   
+
+            # start with basic list
+            for system in system_names:
+                formatted_list.append({'name': system["system_name"], 'type': 'bar', 'x': [], 'y': []})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:488]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 12]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+            risk_renewal_formatted_list = []
+            total_sum = 0
+            for k in list(risk_renewal_of_assets):
+                if k["values"] is not None:
+                    values = k["values"]
+                else:
+                    values = 0
+
+                if k not in risk_renewal_formatted_list:
+                    risk_renewal_formatted_list.append({'system_name': k["system_name"], "risk": [], "total_value": 0, "total_per": 0})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:515]: same with above

File: django_project/civitas/models/view/reporter_data.py [hunk 13]
```diff
@@ -299,50 +309,306 @@ def dashboard(community: Community) -> dict:
...
+                        else:
+                            _values = 0
+
+                        key["risk"].append({"risk_level": value, "values": _values})
+                        key["total_value"] = float(key["total_value"]) + float(_values)
+                        key["total_per"] = (float(key["total_value"] ) / total_sum) * 100
+
+            for key in final_risk_renewal:
+                if not any(d['risk_level'] == 'Extreme' for d in key["risk"]):
+                    key["risk"].append({"risk_level": 'Extreme', "values": 0.00})
```
Comment (meomancer) [inline django_project/civitas/models/view/reporter_data.py:540]: same with above

Comment (meomancer) [reply django_project/civitas/models/view/reporter_data.py:540]: same with below
Please make the append newline

File: django_project/web-app/static/js/dashboard/community/controller.js [hunk 14]
```diff
@@ -94,5 +95,25 @@ define([
             }, new maplibregl.LngLatBounds(coordinates[0], coordinates[0]));
             event.trigger(evt.MAP_FLY, bounds);
         },
+        featureChanged: function (geojson) {
+            if (!geojson?.geometry?.coordinates || !geojson?.geometry?.coordinates[0]) {
+                return
+            }
+            console.log("feature new", geojson)
```
Comment (meomancer) [inline django_project/web-app/static/js/dashboard/community/controller.js:102]: Please delete it

**Comment by voogt** [comment]
Hi @meomancer, I've moved the logic for formatting the data to the file called dashboard_data.py and updated the syntax as requested there.

**Pull request merged**